src/operators/consolidate.rs

```diff
@@ -106,3 +106,157 @@ where
+    use timely::dataflow::channels::pushers::buffer::Session;
+    use timely::dataflow::channels::pushers::{Counter, Tee};
+    use timely::dataflow::operators::{Capability, Operator};
+    use timely::progress::{Antichain, Timestamp};
+
+    use crate::collection::AsCollection;
+    use crate::difference::Semigroup;
+    use crate::lattice::Lattice;
+    use crate::trace::{Batcher, Builder};
+    use crate::{Collection, Data, ExchangeData, Hashable};
+
+    impl<G, D, R> Collection<G, D, R>
+        where
+            G: Scope,
+            G::Timestamp: Data + Lattice,
+            D: ExchangeData + Hashable,
+            R: Semigroup + ExchangeData,
+    {
+        /// Aggregates the weights of equal records into at most one record.
+        ///
+        /// This method uses the type `D`'s `hashed()` method to partition the data. The data are
+        /// accumulated in place, each held back until their timestamp has completed.
+        ///
+        /// # Examples
+        ///
+        /// ```
+        /// use differential_dataflow::input::Input;
+        /// use differential_dataflow::trace::implementations::merge_batcher::MergeBatcher;
+        ///
+        /// ::timely::example(|scope| {
+        ///
+        ///     let x = scope.new_collection_from(1 .. 10u32).1;
+        ///
+        ///     x.negate()
+        ///      .concat(&x)
+        ///      .consolidate_named_neu::<MergeBatcher<_, _, _, _>>("Consolidate inputs") // <-- ensures cancellation occurs
+        ///      .assert_empty();
+        /// });
+        /// ```
+        pub fn consolidate_named_neu<B>(&self, name: &str) -> Self
+            where
+                B: Batcher<Item=((D, ()), G::Timestamp, R), Time=G::Timestamp> + 'static,
+        {
+            let exchange = Exchange::new(move |update: &((D, ()), G::Timestamp, R)| (update.0).0.hashed().into());
+            self.map(|k| (k, ())).inner
+                .unary_frontier(exchange, name, |_cap, info| {
+
+                    // Acquire a logger for arrange events.
+                    let logger = {
+                        let scope = self.scope();
+                        let register = scope.log_register();
+                        register.get::<crate::logging::DifferentialEvent>("differential/arrange")
+                    };
+
+                    let mut batcher = B::new(logger, info.global_id);
+                    // Capabilities for the lower envelope of updates in `batcher`.
+                    let mut capabilities = Antichain::<Capability<G::Timestamp>>::new();
+                    let mut prev_frontier = Antichain::from_elem(<G::Timestamp as Timestamp>::minimum());
+
+
+                    move |input, output| {
+                        input.for_each(|cap, data| {
+                            capabilities.insert(cap.retain());
+                            batcher.push_batch(data);
+                        });
+
+                        if prev_frontier.borrow() != input.frontier().frontier() {
```

Comment:
I stole the capability management logic from arrange_core, so it might check out?